clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -538,6 +542,58 @@ bindConstantVar = inlineBinders test
         _ -> return False
     -- test _ _ = return False
 
+-- | Only inline casts that just contain a 'Var', because these are guaranteed work-free.
+-- These are the result of the 'splitCast' transformation.
+inlineCast :: NormRewrite
+inlineCast = inlineBinders test
+  where
+    test _ (_, Embed (Cast (Var _ _) _ _)) = return True
+    test _ _ = return False
+
+-- | Eleminate two back to back casts where the type going in and coming out are the same
+--
+-- @
+--   (cast :: b -> a) $ (cast :: a -> b) x   ==> x
+-- @
+eliminateCastCast :: NormRewrite
```

Comment:
It's good to check tyA and tyB.
But we need to do type normalization before comparing them, because they could be for example: `Signed 7` and `Signed (6+1)`.

I'm changing the case where tyA and tyC don't match, to produce an error.
Introducing a new cast here will produce a cast that can never be eliminated.
And then we will get an error during netlist generation that there was still some cast left over, but at that point we won't know where it came from.

Reply:
Also in the test case that I have this doesn't actually provide any source location.